FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/Jobs/SchedulerService.cs

```diff
@@ -64,188 +73,260 @@ public class SchedulerService : ISchedulerService
 
         public int SchedulerServiceLeaseRefreshIntervalInSeconds { get; set; } = JobConfigurationConstants.DefaultSchedulerServiceLeaseRefreshIntervalInSeconds;
 
+        // scheduler service is a long running service, it shouldn't stop for any exception.
+        // It stops only when the job is scheduled to end or is cancelled.
         public async Task RunAsync(CancellationToken cancellationToken)
         {
-            _diagnosticLogger.LogInformation("Scheduler starts running.");
-            _logger.LogInformation("Scheduler starts running.");
-            using CancellationTokenSource cancellationTokenSource = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
-            CancellationToken serviceCancellationToken = cancellationTokenSource.Token;
-            var stopRunning = false;
+            _diagnosticLogger.LogInformation($"Scheduler instance {_instanceGuid} starts running.");
+            _logger.LogInformation($"Scheduler instance {_instanceGuid} starts running.");
 
-            while (!stopRunning && !serviceCancellationToken.IsCancellationRequested)
+            bool scheduledToEnd = false;
+            while (!scheduledToEnd && !cancellationToken.IsCancellationRequested)
             {
-                var delayTask = Task.Delay(TimeSpan.FromSeconds(SchedulerServicePullingIntervalInSeconds), CancellationToken.None);
+                // delay task should stop without throwing exception when is cancellation requested.
+                Task delayTask = Task.Delay(TimeSpan.FromSeconds(SchedulerServicePullingIntervalInSeconds), cancellationToken).ContinueWith(_ => { });
 
                 try
                 {
-                    // try next time if the queue client isn't initialized.
+                    // try next time if the storage aren't initialized.
                     if (!_queueClient.IsInitialized() || !_metadataStore.IsInitialized())
```

Comment:
Does those two initializations also need lock to make them thread-safety?

Reply:
Per our offline discussion, the design is ok.